alf/algorithms/mcts_algorithm.py

```diff
@@ -338,9 +379,35 @@ def predict_step(self, time_step: TimeStep, state: MCTSState):
         # will be based on none-zero ucb_scores
         trees.visit_count[roots] = 1
         trees.value_sum[roots] = model_output.value
-        self._update_best_child(trees, roots)
+        if not self._expand_all_children:
+            self._build_tree1(trees, to_plays)
+        else:
+            self._build_tree2(trees, to_plays)
 
-        for sim in range(1, self._num_simulations + 1):
+        action, info = self._select_action(trees, steps)
+        return AlgStep(
+            output=action, state=MCTSState(steps=state.steps + 1), info=info)
+
+    def _build_tree1(self, trees, to_plays):
+        """Build the tree by evaluating one node a time"""
+        roots = (trees.B, trees.root_indices)
+        sim0 = 1
+        if self._expand_all_root_children:
+            v = self._expand_children(trees, roots, 1, to_plays)
+            branch_factor = trees.branch_factor
+            trees.visit_count[roots] += branch_factor
+            trees.value_sum[roots] += branch_factor * v
+            children = trees.get_children(roots)
+            nodes = torch.broadcast_tensors(*roots)
```

Comment:
I do not understand immediately why we need to do `broadcast_tensors`. If `roots` is something like 

```python
(tensor([0, 1, 2, 3]), tensor([0, 0, 0, 0]))
```

Assuming 4 batches. How is `nodes` different from `roots` after applying the `broadcast_tensor`?